### PR TITLE
tests/basic.t: add more test cases

### PR DESCRIPTION
**Pull request by lxbsz**
Problem:
-------
Need to more test cases to do the code coverage.

Solution:
--------
Add more test cases that possiblly, such as for the 'create' add
'ha|prealloc|ring-buffer....', etc.

Signed-off-by: Xiubo Li <xiubli@redhat.com>

### Does this PR fix issues?

Fixes: #

### CODE REVIEW

File: tests/gfapi-test.c [hunk 1]
```diff
@@ -89,8 +89,8 @@ main(int argc, char *argv[])
 
 out:
   if (tgfd) {
-    glfs_unlink(fs, GB_TEST_FILE);
     glfs_close(tgfd);
+    glfs_unlink(fs, GB_TEST_FILE);
```
Comment (amarts) [inline tests/gfapi-test.c]: thinking more about this, should this line move out of 'if()' condition check?

Comment (lxbsz) [reply tests/gfapi-test.c]: IMO, No need and logic will be like:
```
 
tgfd = create();
if (!tgfd) {
     ....
    goto out;
}

tgfd = open();
if (!tgfd) {
    ....
    glfs_unlike();
    goto out;
}

out:
if (tgfd) {
    glfs_close();
    glfs_unlike();
}
```

File: tests/basic.t [hunk 2]
```diff
@@ -19,7 +19,7 @@
 HOST=$(hostname -I | awk '{print $1}')
 VOLNAME="hosting-volume"
 BLKNAME="block-volume"
-BRKDIR="/brick"
+BRKDIR="/data/brick"
```
Comment (pkalever) [inline tests/basic.t]: You should 'mkdir /data/' somewhere 
Else volume create will fail IMO.

Comment (lxbsz) [reply tests/basic.t]: Yeah, will fix it.
Thanks.

**Comment by lxbsz** [comment]
@pkalever @amarts 
Updated done.
Thanks.

File: tests/basic.t [hunk 3]
```diff
@@ -65,49 +65,94 @@ if [ $? -eq 1 ]; then
...
 TEST ./tests/gfapi-test ${VOLNAME} ${HOST}
+##### End #####
 
-# Start gluster-blockd.service
+
+##### Restart gluster-blockd.service Test Start #####
 systemctl daemon-reload
-TEST systemctl restart gluster-blockd.service
+TEST systemctl restart gluster-block-target.service gluster-blockd.service
 sleep 1;
```
Comment (pkalever) [inline tests/basic.t]: @lxbsz , lets make this sleep time to 5 seconds.

I see:
TEST : systemctl restart gluster-block-target.service gluster-blockd.service
line 83 : OK                                                           
                                                                      
TEST : gluster-block create hosting-volume/block-volume 192.168.124.147 1MiB                                          
client create failed: RPC: Remote system error - Connection refused, unix addr /var/run/gluster-blockd.socket         
line 90 : NOT OK

Lets give daemon some more time to finish the init stuff

File: tests/basic.t [hunk 4]
```diff
@@ -65,49 +65,94 @@ if [ $? -eq 1 ]; then
...
+# Block create with 'ha' set/delete
+TEST gluster-block create ${VOLNAME}/${BLKNAME} ha 1 ${HOST} 1MiB
+TEST gluster-block delete ${VOLNAME}/${BLKNAME}
+
+# Block create with 'prealloc' set/delete
+TEST gluster-block create ${VOLNAME}/${BLKNAME} prealloc full ${HOST} 1MiB
+TEST gluster-block delete ${VOLNAME}/${BLKNAME}
+
+# Block create with 'ring-buffer' set/delete
+TEST gluster-block create ${VOLNAME}/${BLKNAME} ring-buffer 32 ${HOST} 1MiB
```
Comment (pkalever) [inline tests/basic.t]: @lxbsz This is failing, can you check:

TEST : gluster-block create hosting-volume/block-volume ring-buffer 32 192.168.124.147 1MiB
failed to configure on 192.168.124.147 configure failed                                                  
RESULT:FAIL                                                                                                
line 108 : NOT OK

File: tests/basic.t (hunk 3 above)
Comment (lxbsz) [reply tests/basic.t]: Is 3 seconds okay? or 5 ?

Comment (pkalever) [reply tests/basic.t]: Its okay with 5 seconds wait. These are just tests, lets give sometime for it.

File: tests/basic.t (hunk 4 above)
Comment (lxbsz) [reply tests/basic.t]: What's the kernel version ?
My local test is okay in RHEL and fedora 29.

Comment (pkalever) [reply tests/basic.t]: [root@Fedora-Server gluster-block]# uname -a
Linux Fedora-Server 5.0.3-200.fc29.x86_64 #1 SMP Tue Mar 19 15:07:58 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
[root@Fedora-Server gluster-block]# cat /etc/fedora-release 
Fedora release 29 (Twenty Nine)
[root@Fedora-Server gluster-block]# 



@lxbsz may be its coming from rtslib-v2.1.fb69 ? 

If it is passing for you ignore, I will debug the issue locally.

Thanks!

Comment (lxbsz) [reply tests/basic.t]: yeah, sure please, and at the same time I will test more.

**Comment by pkalever** [comment]
@lxbsz please update this PR with 5 sec wait. Post that will merge it. Thanks!
Reviewed-by: Prasanna Kumar Kalever \<prasanna.kalever@redhat.com\>

**Comment by lxbsz** [comment]
> @lxbsz please update this PR with 5 sec wait. Post that will merge it. Thanks!
> Reviewed-by: Prasanna Kumar Kalever <[prasanna.kalever@redhat.com](mailto:prasanna.kalever@redhat.com)>

Update done.

File: tests/basic.t [hunk 5]
```diff
@@ -65,49 +65,94 @@ if [ $? -eq 1 ]; then
...
 TEST ./tests/gfapi-test ${VOLNAME} ${HOST}
+##### End #####
 
-# Start gluster-blockd.service
+
+##### Restart gluster-blockd.service Test Start #####
 systemctl daemon-reload
-TEST systemctl restart gluster-blockd.service
-sleep 1;
+TEST systemctl restart gluster-block-target.service gluster-blockd.service
```
Comment (pkalever) [inline tests/basic.t]: Why do we need to specify gluster-block-target.service ?

File: tests/basic.t [hunk 6]
```diff
@@ -65,49 +65,94 @@ if [ $? -eq 1 ]; then
...
+TEST gluster-block create ${VOLNAME}/${BLKNAME} auth enable ${HOST} 1MiB
+TEST gluster-block delete ${VOLNAME}/${BLKNAME}
+
+# Block create with 'ha' set/delete
+TEST gluster-block create ${VOLNAME}/${BLKNAME} ha 1 ${HOST} 1MiB
+TEST gluster-block delete ${VOLNAME}/${BLKNAME}
+
+# Block create with 'prealloc' set/delete
+TEST gluster-block create ${VOLNAME}/${BLKNAME} prealloc full ${HOST} 1MiB
+TEST gluster-block delete ${VOLNAME}/${BLKNAME}
```
Comment (pkalever) [inline tests/basic.t]: as prealloc is full now by default, should we also test for prealloc no ?

File: tests/basic.t (hunk 5 above)
Comment (lxbsz) [reply tests/basic.t]: I am okay to remove this.
I thought it also part of the gluster-block project.

File: tests/basic.t (hunk 6 above)
Comment (lxbsz) [reply tests/basic.t]: Ah, yeah, right, I forgot that changes.
Will add this.

File: tests/basic.t (hunk 5 above)
Comment (pkalever) [reply tests/basic.t]: Yeah, gluster-block should internally take care of triggering gluster-block-target, so no need to explicitly mention it here.

Thanks!

Comment (lxbsz) [reply tests/basic.t]: yeah, sure, will remove it.

**Comment by lxbsz** [comment]
@pkalever 
Update it.

**Comment by pkalever** [comment]
> @lxbsz one last review comments:
> 
> * This changes miss to add test for create with 'storage' option and delete with 'unlink' option.
> * prealloc no test
> * genconfig test
> * timeout option test
> 
> Is this something that we can accommodate here ?
> 
> Post that we can work on negative test cases as part of a separate PR.
> 
@lxbsz thank you! but looks like you have missed this comment.

**Comment by lxbsz** [comment]
> > @lxbsz one last review comments:
> > 
> > * This changes miss to add test for create with 'storage' option and delete with 'unlink' option.
> > * prealloc no test
> > * genconfig test
> > * timeout option test
> > 
> > Is this something that we can accommodate here ?
> > Post that we can work on negative test cases as part of a separate PR.
> 
> @lxbsz thank you! but looks like you have missed this comment.

Yeah, doing it now.

**Comment by lxbsz** [comment]
@pkalever 
Update it and please review again.
Thanks

File: tests/basic.t [hunk 7]
```diff
@@ -65,49 +66,125 @@ if [ $? -eq 1 ]; then
...
+TEST gluster-block create ${VOLNAME}/${BLKNAME} prealloc no ${HOST} 1MiB
+TEST gluster-block delete ${VOLNAME}/${BLKNAME}
+
+# Block create with 'ring-buffer' set/delete
+TEST gluster-block create ${VOLNAME}/${BLKNAME} ring-buffer 32 ${HOST} 1MiB
+TEST gluster-block delete ${VOLNAME}/${BLKNAME}
+
+# Block create with 'storage' set and delete with 'unlink-storage no' set
+TEST gluster-block create ${VOLNAME}/${BLKNAME} ${HOST} 1MiB
+LINK=`eval gluster-block info ${VOLNAME}/${BLKNAME} | grep GBID | awk -F: '{print $2}'`
```
Comment (pkalever) [inline tests/basic.t]: The delimiter should be -F' ' (space)

Comment (pkalever) [reply tests/basic.t]: May be awk -F' ' '{print $NF}' is better

Comment (lxbsz) [reply tests/basic.t]: make sense.

**Comment by lxbsz** [comment]
@pkalever 
Update done.
Thanks.

**Pull request merged**